concordium-consensus/src/Concordium/GlobalState/Persistent/Account/StructureV1.hs

```diff
@@ -0,0 +1,1314 @@
+    load = do
+        paedHash <- get
+        EnduringDataFlags{..} <- get
+        mPersistingData <- load
+        mEncryptedAmount <-
+            if edHasEncryptedAmount
+                then load
+                else return (return Null)
+        mReleaseSchedule <- if edHasReleaseSchedule then load else return (return Null)
+        let getPC PendingChangeNone = return NoChange
+            getPC PendingChangeReduce = ReduceStake <$> get <*> get
+            getPC PendingChangeRemove = RemoveStake <$> get
+        mStake <- case edStakeFlags of
+            StakeFlagsNone -> return (return PersistentAccountStakeEnduringNone)
+            StakeFlagsBaker{..} -> do
+                let paseBakerRestakeEarnings = sfRestake
+                mInfo <- load
+                paseBakerPendingChange <- getPC sfChangeType
+                return $! do
+                    paseBakerInfo <- mInfo
+                    return PersistentAccountStakeEnduringBaker{..}
+            StakeFlagsDelegator{..} -> do
+                paseDelegatorId <- get
+                let paseDelegatorRestakeEarnings = sfRestake
+                paseDelegatorTarget <- if sfPassive then return DelegatePassive else DelegateToBaker <$> get
+                paseDelegatorPendingChange <- getPC sfChangeType
+                return . return $! PersistentAccountStakeEnduringDelegator{..}
+        return $! do
+            paedPersistingData <- mPersistingData
+            paedEncryptedAmount <- mEncryptedAmount
+            paedReleaseSchedule <- mReleaseSchedule
+            paedStake <- mStake
+            return PersistentAccountEnduringData{..}
+
+-- * Persistent account
+
+-- |A persistent account.
+-- The most commonly modified fields of an account (the next nonce, balance and staked balance)
+-- are directly available.  The rest of the fields are stored as part of the enduring data,
+-- under an 'EagerBufferedRef'.  This limits the amount that needs to be rewritten for the
+-- most common updates.
+data PersistentAccount av = PersistentAccount
+    { -- |The next nonce for transactions on the account.
+      accountNonce :: !Nonce,
+      -- |The total balance of the account.
+      accountAmount :: !Amount,
+      -- |The staked balance of the account.
+      -- INVARIANT: This is 0 if the account is not a baker or delegator.
+      accountStakedAmount :: !Amount,
+      -- |The enduring account data.
+      accountEnduringData :: !(EagerBufferedRef (PersistentAccountEnduringData av))
+    }
+
+instance HashableTo (AccountHash 'AccountV2) (PersistentAccount 'AccountV2) where
+    getHash PersistentAccount{..} =
+        makeAccountHash $
+            AHIV2 $
+                AccountHashInputsV2
+                    { ahi2NextNonce = accountNonce,
+                      ahi2AccountBalance = accountAmount,
+                      ahi2StakedBalance = accountStakedAmount,
+                      ahi2MerkleHash = getHash accountEnduringData
```

Comment:
If I understand correctly this will load the entire enduring data to compute the hash (which is then just looked up).

It would be better to not have to load all that data and that seems possible since the storage format puts the hash at the start.

Reply:
That is not correct. The enduring data is stored under an `EagerBufferedRef`, which means that it is always loaded when the account is loaded.